src/ut_pex.cpp

```diff
@@ -667,46 +653,47 @@ namespace libtorrent { namespace
 		if (pc.type() != peer_connection::bittorrent_connection)
 			return boost::shared_ptr<peer_plugin>();
 
-		return boost::shared_ptr<peer_plugin>(new ut_pex_peer_plugin(m_torrent
-			, *pc.native_handle(), *this));
+		bt_peer_connection* c = static_cast<bt_peer_connection*>(pc.native_handle().get());
+		auto p = boost::make_shared<ut_pex_peer_plugin>(ut_pex_peer_plugin(m_torrent, *c, *this));
+		c->set_ut_pex(p);
+		return p;
 	}
 } }
 
 namespace libtorrent
 {
-	boost::shared_ptr<torrent_plugin> create_ut_pex_plugin(torrent_handle const& th, void*)
-	{
-		torrent* t = th.native_handle().get();
-		if (t->torrent_file().priv() || (t->torrent_file().is_i2p()
-			&& !t->settings().get_bool(settings_pack::allow_i2p_mixed)))
-		{
-			return boost::shared_ptr<torrent_plugin>();
-		}
-		return boost::shared_ptr<torrent_plugin>(new ut_pex_plugin(*t));
-	}
-
-	bool was_introduced_by(peer_plugin const* pp, tcp::endpoint const& ep)
+	bool ut_pex_peer_store::was_introduced_by(tcp::endpoint const &ep)
 	{
-		ut_pex_peer_plugin const* p = static_cast<ut_pex_peer_plugin const*>(pp);
 #if TORRENT_USE_IPV6
 		if (ep.address().is_v4())
 		{
 #endif
-			ut_pex_peer_plugin::peers4_t::value_type v(ep.address().to_v4().to_bytes(), ep.port());
-			ut_pex_peer_plugin::peers4_t::const_iterator i
-				= std::lower_bound(p->m_peers.begin(), p->m_peers.end(), v);
-			return i != p->m_peers.end() && *i == v;
+			peers4_t::value_type v(ep.address().to_v4().to_bytes(), ep.port());
+			peers4_t::const_iterator i
+				= std::lower_bound(m_peers.begin(), m_peers.end(), v);
+			return i != m_peers.end() && *i == v;
 #if TORRENT_USE_IPV6
 		}
 		else
 		{
-			ut_pex_peer_plugin::peers6_t::value_type v(ep.address().to_v6().to_bytes(), ep.port());
-			ut_pex_peer_plugin::peers6_t::const_iterator i
-				= std::lower_bound(p->m_peers6.begin(), p->m_peers6.end(), v);
-			return i != p->m_peers6.end() && *i == v;
+			peers6_t::value_type v(ep.address().to_v6().to_bytes(), ep.port());
+			peers6_t::const_iterator i
+				= std::lower_bound(m_peers6.begin(), m_peers6.end(), v);
+			return i != m_peers6.end() && *i == v;
 		}
 #endif
 	}
+
+	boost::shared_ptr<torrent_plugin> create_ut_pex_plugin(torrent_handle const& th, void*)
+	{
+		torrent* t = th.native_handle().get();
+		if (t->torrent_file().priv() || (t->torrent_file().is_i2p()
+			&& !t->settings().get_bool(settings_pack::allow_i2p_mixed)))
+		{
+			return boost::shared_ptr<torrent_plugin>();
+		}
+		return boost::shared_ptr<torrent_plugin>(new ut_pex_plugin(*t));
```

Comment:
I think the same argument applies here. ut_pex_plugin has internal linkage and could have a public constructor, and boost::make_shared<> could be used